src/Microsoft.DotNet.Build.Tasks/Targets/publishtest.targets

```diff
@@ -20,18 +21,11 @@
 
     <Exec Command="$(NugetRestoreCommand) &quot;$(TestRuntimePackageConfig)&quot;" StandardOutputImportance="Low" />
 
-    <!-- DNU will emit a project.lock.json.  When multiple tests are building for the first time they may race on 
-         running this target and generating the lock.json.  This is causing IOExceptions in dnu for writing lock.json.
-         We should really be handling these test dependencies in a different way (as a package themselves) but for now
-         just restore the project.json from intermediate to avoid the race.  -->
-    <Copy SourceFiles="$(TestRuntimeProjectJson)" DestinationFolder="$(IntermediateOutputPath)">
-      <Output TaskParameter="CopiedFiles" PropertyName="TempTestRuntimeProjectJson"/>
-    </Copy>
-    <Exec Command="$(DnuRestoreCommand) &quot;$(TempTestRuntimeProjectJson)&quot;" StandardOutputImportance="Low" CustomErrorRegularExpression="^Unable to locate .*" />
+    <ExecWithMutex Command="$(DnuRestoreCommand) &quot;$(TestRuntimeProjectJson)&quot;" MutexName="$(TestRuntimeProjectJson)" StandardOutputImportance="Low" CustomErrorRegularExpression="^Unable to locate .*" />
```

Comment:
The only time this comes into play is the very first build.  The restore is incremental so as soon as it succeeds one no further projects will run it.  This effectively means that worst case N-1 (where n is the number of MSBuild nodes) projects will be waiting once / serialized while this restore happens, after that this code will not be hit.

I discussed the issue with the nuget/dnx team and they said this is effectively what they plan to do there to fix the bug.